File: Sample-ObjC/ViewController.m
//
//  ViewController.m
//  ObjcSample
//
//  Created by artur on 06.05.2020.
//  Copyright © 2020 artur. All rights reserved.
//

#import "ViewController.h"
#import "AppDelegate.h"

#import <HolisticSolutionSDK/HolisticSolutionSDK.h>
#import <Appodeal/Appodeal.h>
#import <Adjust/Adjust.h>
#import <AppsFlyerLib/AppsFlyerLib.h>
#import <Firebase.h>
#import <FirebaseRemoteConfig.h>
#import <FBSDKCoreKit/FBSDKCoreKit.h>


@interface ViewController ()

@property (weak, nonatomic) IBOutlet UILabel *adjustVersion;
@property (weak, nonatomic) IBOutlet UILabel *adjustAttID;
@property (weak, nonatomic) IBOutlet UILabel *appsflyerVersion;
@property (weak, nonatomic) IBOutlet UILabel *appsflyerAttID;
@property (weak, nonatomic) IBOutlet UILabel *firebaseVersion;
@property (weak, nonatomic) IBOutlet UILabel *firebaseKeyWords;
@property (weak, nonatomic) IBOutlet UILabel *facebookVersion;
@property (weak, nonatomic) IBOutlet UILabel *facebookAppID;
@property (weak, nonatomic) IBOutlet UILabel *apdVersion;
@property (weak, nonatomic) IBOutlet UILabel *apdInitialized;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(didInitialiseAd)
                                                 name:completeNotification
                                               object:nil];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)didInitialiseAd {
    [Appodeal showAd:AppodealShowStyleBannerBottom rootViewController:self];
    [self updateLabels];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    NSString *selected = [tableView cellForRowAtIndexPath:indexPath].textLabel.text;
    if ([selected isEqualToString:@"Event"]) {
        [self synthesizeEvent];
    } else if ([selected isEqualToString:@"Purchase"]) {
        [self synthesizePurchase:HSPurchaseTypeConsumable];
    } else if ([selected isEqualToString:@"Subscription"]) {
        [self synthesizePurchase:HSPurchaseTypeAutoRenewableSubscription];
    } else {
        return;
    }
}

- (void)synthesizePurchase:(HSPurchaseType)type {
    NSDictionary *params = @{
        @"Test Custom 1" : @"Value 1",
        @"Test Custom 2" : @"Value 2"
    };
    __weak typeof(self) weakSelf = self;
    [Appodeal.hs validateAndTrackInAppPurchaseWithProductId:@"some product id"
                                                       type:type
                                                      price:@"9.99"
                                                   currency:@"USD"
                                              transactionId:@"some transiton id"
                                       additionalParameters:params
                                                    success:^(NSDictionary *result) {
        [weakSelf alertWithTitle:@"Purchase is valid" message:result.description];
    }
                                                    failure:^(NSError *error, id obj) {
        [weakSelf alertWithTitle:@"Purchase is invalid" message:error.localizedDescription];
    }];
}

- (void)synthesizeEvent {
    [Appodeal.hs trackEvent:@"level_started" customParameters:nil];
}

- (void)alertWithTitle:(NSString *)title message:(NSString * _Nullable)message {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title
                                                                   message:message
                                                            preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleCancel handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)updateLabels {
    //adjust
    _adjustVersion.text = Adjust.sdkVersion;
    _adjustAttID.text = Adjust.adid;
    
    //appsflyer
    _appsflyerVersion.text = AppsFlyerLib.shared.getSDKVersion;
    _appsflyerAttID.text = AppsFlyerLib.shared.getAppsFlyerUID;
    
    //firebase
    _firebaseVersion.text = FIRFirebaseVersion();
    _firebaseKeyWords.text = [[FIRRemoteConfig.remoteConfig allKeysFromSource:FIRRemoteConfigSourceRemote] componentsJoinedByString:@", "];
    
    //facebook
    _facebookVersion.text = FBSDK_VERSION_STRING;
    _facebookAppID.text = [NSBundle.mainBundle objectForInfoDictionaryKey:@"FacebookAppID"];
    
    //appodeal
    _apdVersion.text = APDSdkVersionString();
    _apdInitialized.text = [Appodeal isInitalizedForAdType:kAppodealTypes] ? @"true" : @"false";
}

@end

File: Sample-Swift/ViewController.swift
//
//  ViewController.swift
//  Sample
//
//  Created by Stas Kochkin on 04.05.2020.
//  Copyright © 2020 com.appodeal. All rights reserved.
//

import UIKit
import Appodeal
import HolisticSolutionSDK
import Adjust
import AppsFlyerLib
import Firebase
import FirebaseRemoteConfig
import FBSDKCoreKit


private extension UITableViewCell {
    static func base(
        title: String,
        detail: String? = nil,
        style: UITableViewCell.CellStyle = .value1,
        accesory: UITableViewCell.AccessoryType = .none,
        userInteractionEnabled: Bool = false,
        reuseIdentifier: String = ViewController.cellId
    ) -> UITableViewCell {
        let cell = UITableViewCell(style: style, reuseIdentifier: reuseIdentifier)
        cell.isUserInteractionEnabled = userInteractionEnabled
        cell.textLabel?.text = title
        cell.detailTextLabel?.text = detail
        cell.accessoryType = accesory
        return cell
    }
}

final class ViewController: StaticTableViewController {
    fileprivate static let cellId: String = "basic.cell"
    
    override init() {
        super.init()
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(holisitcSolutionDidComplete),
            name: AppDelegate.complete,
            object: nil
        )
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "basic.cell")
        sections = [
            behaviour,
            adjust,
            appsflyer,
            firebase,
            facebook,
            appodeal
        ]
    }
    
    private var adjust: Section {
        .init(
            header: "Adjust",
            rows: [
                .init { _ in .base(title: "Version", detail: Adjust.sdkVersion()) },
                .init { _ in .base(title: "Attribution ID", detail: Adjust.adid() ?? "-") }
            ]
        )
    }
    
    private var appsflyer: Section {
        .init(
            header: "AppsFlyer",
            rows: [
                .init { _ in .base(title: "Version", detail: AppsFlyerLib.shared().getSDKVersion()) },
                .init { _ in .base(title: "Attribution ID", detail:  AppsFlyerLib.shared().getAppsFlyerUID()) }
            ]
        )
    }
    
    private var firebase: Section {
        .init(
            header: "Firebase (Remote Config)",
            rows: [
                .init { _ in .base(title: "Version", detail: FirebaseVersion()) },
                .init { _ in .base(title: "Keywords", detail: RemoteConfig.remoteConfig().allKeys(from: .remote).joined(separator: ", ")) }
            ]
        )
    }
    
    private var facebook: Section {
        .init(
            header: "FB SDK",
            rows: [
                .init { _ in .base(title: "Version", detail: FBSDK_VERSION_STRING) },
                .init { _ in .base(title: "App ID", detail: Bundle.main.object(forInfoDictionaryKey: "FacebookAppID") as? String) }
            ]
        )
    }
    
    
    private var appodeal: Section {
        .init(
            header: "Appodeal",
            rows: [
                .init { _ in .base(title: "Version", detail: APDSdkVersionString()) },
                .init { _ in .base(title: "Initialized", accesory: Appodeal.hs.initialized ? .checkmark : .none) },
            ]
        )
    }
    
    private var behaviour: Section {
        .init(
            header: "Behavior",
            rows: [
                .init(
                    cell: { _ in .base(title: "Event", detail: "Synthesize random event", style: .subtitle, userInteractionEnabled: true) },
                    select: { [unowned self] in self.synthesizeEvent() }
                ),
                .init(
                    cell: { _ in .base(title: "Purchase", detail: "Synthesize random in-app purchase", style: .subtitle, userInteractionEnabled: true) },
                    select: { [unowned self] in self.synthesizePurchase(.consumable) }
                ),
                .init(
                    cell: { _ in .base(title: "Subscription", detail: "Synthesize random subscription", style: .subtitle, userInteractionEnabled: true) },
                    select: { [unowned self] in self.synthesizePurchase(.autoRenewableSubscription) }
                )
            ]
        )
    }
    
    
    @objc private
    func holisitcSolutionDidComplete() {
        Appodeal.showAd(.bannerBottom, rootViewController: self)
        tableView.reloadSections(IndexSet(1..<sections.count), with: .fade)
    }
    
    func synthesizePurchase(_ type: PurchaseType) {
        let additionalParameters = [
            "Test Custom 1" : "Value 1",
            "Test Custom 2" : "Value 2"
        ]
        Appodeal.hs.validateAndTrackInAppPurchase(
            productId: "some product id",
            type: type,
            price: "9.99",
            currency: "USD",
            transactionId: "some transaction id",
            additionalParameters: additionalParameters,
            success: { [weak self] in self?.alert("Purchase is valid", message: $0.description) },
            failure: { [weak self] error, _ in self?.alert("Purchase is invalid", message: error?.localizedDescription) }
        )
    }
    
    func synthesizeEvent() {
        Appodeal.hs.trackEvent("start_game", customParameters: nil)
    }
    
    func alert(_ title: String, message: String? = nil) {
        let alert = UIAlertController(
            title: title,
            message: message,
            preferredStyle: .alert
        )
        alert.addAction(UIAlertAction(title: "OK", style: .cancel, handler: nil))
        present(alert, animated: true, completion: nil)
    }
}

